### ISSUE TITLE
[FDSP2] express zero-1 with fully_shard

### ISSUE THREAD

**Issue by weifengpy** [pytorch/pytorch#155952]
### 🚀 The feature, motivation and pitch

zero-1 is still interesting for recommedation model. sharding optimizer states might be enough to reduce memory footprint

torch/distributed/optim/zero_redundancy_optimizer.py is not maintained. Another opion is to express zero-1 with fully_shard
* model.parameters as DTensor for constructing optimizer
* model.state_dict as plain tensor
* no all-gather, no reduce-scatter, just all-reduce
* pre-forward/backward, switch model.paramters() from DTensor to plain tensor
* post-forward/backward, switch model.parameters() from plain tensor to DTensor

<img width="814" alt="Image" src="https://github.com/user-attachments/assets/f58a5d71-ebc7-4ffc-9a5f-be66cceb4982" />

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

**Comment by weifengpy** [comment]
cc @awgu for thoughts on expressing zero-1 with fully_shard

**Comment by wanchaol** [comment]
🤔 ZeRO1 does not have anything to do with the model (nn.Modules) though, it only shard the optimizer states, I am not sure it's actually proper to make it a feature in FSDP2, as fully shard API applies to the model not optimizers. IMO the zero_redundancy_optimizer for ZERO1 looks a legit API.

**Comment by weifengpy** [comment]
> 🤔 ZeRO1 does not have anything to do with the model (nn.Modules) though, it only shard the optimizer states, I am not sure it's actually proper to make it a feature in FSDP2, as fully shard API applies to the model not optimizers. IMO the zero_redundancy_optimizer for ZERO1 looks a legit API.

@wanchaol thanks for explaining this. agree that API-wise, zero1 should happen in optimizer. will take a look at zero_redundancy_optimizer seriously

**Comment by weifengpy** [comment]
summarizing discussion with @awgu , for zero-1 implementation, we need several things
* overlapping gradient all-reduce with compute, we need DDP gradient bucketting or module wrapper (fully_shard, or replicate API) for this pupose
* sync parameters after opt.step. This can happen in zero_redundancy_optimizer but we need to verify overlapping as there is limited compute in optimizer. Another option is to all-gather parameters during forward, that gives more room for overlapping. It becomes zero-2 (fully_shard(reshard_after_forward=False)

**Comment by weifengpy** [comment]
not planned
